docs/synapse/userguides/storm_adv_functions.rstorm

```diff
@@ -0,0 +1,708 @@
+  The ability of data emitters to emit data incrementally is useful when consuming large result sets from
+  an API. "Subsets" of results (such as individual JSON objects from a JSON blob) can be made available
+  more quickly (e.g., to another function responsible for creating nodes from the JSON) while the
+  emitter continues to process data.
+  
+  In contrast, if the same set of API results was consumed by a callable function, the function would
+  need to consume the entire result set before returning.
+
+Pseudocode
+**********
+
+::
+  
+  function data_emitter() {
+      
+      for $thing in $things {
+          <do stuff>
+          emit $thing
+      }
+  }
+
+Or:
+
+::
+  
+  function data_emitter() {
+      
+      for $thing in $things {
+          <do stuff>
+          emit $thing
+          
+          if ($thing = "badthing") {
+              stop
+          }
+      }
+  }
+
+Or:
+
+::
+  
+  function data_emitter() {
+  
+      while (1) {
+        <do stuff>
+        emit $thing
+        
+        if (<end condition>) { stop }
+      }
+      
+      <update something to contine while loop>
+  }
+
+
+Example
+*******
+
+Some data sources may paginate results, returning X number of objects (e.g., in a JSON blob) at a time
+until all results are returned. A data emitter function can emit individual JSON objects from the blob
+(e.g., for consumption by another function that processes the object and creates nodes) until all of
+the results have been received.
+
+::
+  
+  function emitReportFeed() {
+      
+      //Set variables for the current time and the # of objects to retrieve per page
+      $now = $lib.time.now()
+      $pagelim = 100
+      
+      //Set a variable for API query parameters
+      $params = ({
+          "limit": $pagelim,
+      })
+      
+      //Set a variable for the initial offset
+      $offset = (0)
+      
+      //While loop to retrieve records
+      while (1) {
+          
+          //Set the value of the 'offset' parameter
+          $params.offset = $offset
+          
+          //Invoke an existing function to retrieve the JSON using $pararms as parameters
+          // to the API request.
+          //Assign the returned JSON to the variable $data
+          $data = $getJson("/reports", params=$params)
+          
+          //If no data is returned, stop and exit this function
+          if ($data = $lib.null) { stop }
+          
+          //If data is returned, loop over the JSON and emit each item / ojbect
+          for $item in $data.data { emit $item }
+          
+          //Set $datasize to the size (number of items) in the returned JSON
+          $datasize = $data.data.size()
+          
+          //Check whether the # of records returned is less than our page limit
+          //If so we have retrieved all available records
+          if ($datasize < $pagelim) {
+                        
+              //Print status to stdout if debug is in use
```

Comment:
Maybe I haven't noticed it before but do we call the CLI output "stdout" elsewhere?

Reply:
changed to CLI